### ISSUE TITLE
attempt to write to file output stream causes "stack smashing detected"

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#588]
Imported from SourceForge on 2024-07-03 02:03:24
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-03-05 23:44:30
Original: https://sourceforge.net/p/maxima/bugs/3291

---

Maxima post-5.39 + ECL 16.1.2 and also ECL 16.1.3 on Ubuntu 14.04. Error not observed with SBCL or Clisp.

Executing `batch("/tmp/t58c.mac", test)` causes `stack smashing detected` and the program dies. The "stack smashing" message is from Glibc or something like that.

Executing `batch("/tmp/t58c.mac")` does not cause an error.

It appears that `openw` succeeds but `printf` causes the crash. I haven't been able to narrow it any further.

Here is the content of /tmp/t58c.mac:

~~~~
output : openw ("/tmp/foo.out");
0;
 
printf (output, "Hello, world!~%");
0;

close (output);
0;

~~~~

Running it with `batch(..., test)`:

~~~~
$ sh maxima-local -l ecl
;;; Loading #P"/usr/local/lib/ecl-16.1.2/sb-bsd-sockets.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/sockets.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/defsystem.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/cmp.fas"
Maxima branch_5_39_base_135_g43a822c_dirty http://maxima.sourceforge.net
using Lisp ECL 16.1.2
Distributed under the GNU Public License. See the file COPYING.
Dedicated to the memory of William Schelter.
The function bug_report() provides bug reporting information.
(%i1) batch("/tmp/t58c.mac", test);

batch: write error log to #<output stream /tmp/t58c.ERR>
********************** Problem 1 (line 1) ***************
Input:
output : openw(/tmp/foo.out)


Result:
#<output stream /tmp/foo.out>

This differed from the expected result:
0
*** stack smashing detected ***: /home/robert/maxima/maxima-code/src/binary-ecl/maxima terminated
Aborted (core dumped)
~~~~

Without `test`:

~~~~
$ sh maxima-local -l ecl
;;; Loading #P"/usr/local/lib/ecl-16.1.2/sb-bsd-sockets.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/sockets.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/defsystem.fas"
;;; Loading #P"/usr/local/lib/ecl-16.1.2/cmp.fas"
Maxima branch_5_39_base_135_g43a822c_dirty http://maxima.sourceforge.net
using Lisp ECL 16.1.2
Distributed under the GNU Public License. See the file COPYING.
Dedicated to the memory of William Schelter.
The function bug_report() provides bug reporting information.
(%i1) batch("/tmp/t58c.mac");

read and interpret file: /tmp/t58c.mac
(%i2) output:openw("/tmp/foo.out")
(%o2)                    #<output stream /tmp/foo.out>
(%i3) 0
(%o3)                                  0
(%i4) printf(output,"Hello, world!~%")
(%o4)                                false
(%i5) 0
(%o5)                                  0
(%i6) close(output)
(%o6)                                true
(%i7) 0
(%o7)                                  0
(%o7)                            /tmp/t58c.mac
(%i8)
~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 02:03:26
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2017-03-06 00:29:04
Original: https://sourceforge.net/p/maxima/bugs/3291/#0039

---

- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -1,4 +1,4 @@
-Maxima post-5.39 + ECL 16.1.2 on Ubuntu 14.04. Error not observed with SBCL or Clisp.
+Maxima post-5.39 + ECL 16.1.2 and also ECL 16.1.3 on Ubuntu 14.04. Error not observed with SBCL or Clisp.
 
 Executing `batch("/tmp/t58c.mac", test)` causes `stack smashing detected` and the program dies. The "stack smashing" message is from Glibc or something like that.
 

~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 02:03:30
Created by **[van_nek](https://sourceforge.net/u/van_nek/)** on 2017-04-29 14:56:44
Original: https://sourceforge.net/p/maxima/bugs/3291/#dcff

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 02:03:33
Created by **[van_nek](https://sourceforge.net/u/van_nek/)** on 2017-04-29 14:56:44
Original: https://sourceforge.net/p/maxima/bugs/3291/#b454

---

Hi Robert,

in 2016 I added the file encoding as an optional argument to openw and friends. If this argument is not given, the encoding is returned from stringproc.lisp / get-encoding. Using ECL the encoding (UTF-8 LF) ist returned, which obviously does not work when test-batching. 

I committed a change to stringproc.lisp and get-encoding now returns just :utf-8 for ECL. That seems to work. 

In the history of stringproc.lisp I see that in October 2016 you changed the default encoding for SBCL in get-encoding. I hope that get-encoding now works as intended in all situations. If not, maybe it is better to cut out this optional argument.

I close this report as fixed.

Best
Volker
